crates/ra_ide_api/src/syntax_highlighting.rs

```diff
@@ -32,59 +33,74 @@ pub(crate) fn highlight(db: &RootDatabase, file_id: FileId) -> Vec<HighlightedRa
-                        Some(Def(ModuleDef::Struct(_))) => "type",
-                        Some(Def(ModuleDef::Union(_))) => "type",
-                        Some(Def(ModuleDef::Enum(_))) => "type",
-                        Some(Def(ModuleDef::EnumVariant(_))) => "constant",
-                        Some(Def(ModuleDef::Const(_))) => "constant",
-                        Some(Def(ModuleDef::Static(_))) => "constant",
-                        Some(Def(ModuleDef::Trait(_))) => "type",
-                        Some(Def(ModuleDef::TypeAlias(_))) => "type",
-                        Some(SelfType(_)) => "type",
-                        Some(Pat(_)) => "text",
-                        Some(SelfParam(_)) => "type",
-                        Some(GenericParam(_)) => "type",
-                        None => "text",
+                        Some(Method(_)) => ("function", None),
+                        Some(Macro(_)) => ("macro", None),
+                        Some(FieldAccess(_)) => ("field", None),
+                        Some(AssocItem(ImplItem::Method(_))) => ("function", None),
+                        Some(AssocItem(ImplItem::Const(_))) => ("constant", None),
+                        Some(AssocItem(ImplItem::TypeAlias(_))) => ("type", None),
+                        Some(Def(ModuleDef::Module(_))) => ("module", None),
+                        Some(Def(ModuleDef::Function(_))) => ("function", None),
+                        Some(Def(ModuleDef::Struct(_))) => ("type", None),
+                        Some(Def(ModuleDef::Union(_))) => ("type", None),
+                        Some(Def(ModuleDef::Enum(_))) => ("type", None),
+                        Some(Def(ModuleDef::EnumVariant(_))) => ("constant", None),
+                        Some(Def(ModuleDef::Const(_))) => ("constant", None),
+                        Some(Def(ModuleDef::Static(_))) => ("constant", None),
+                        Some(Def(ModuleDef::Trait(_))) => ("type", None),
+                        Some(Def(ModuleDef::TypeAlias(_))) => ("type", None),
+                        Some(SelfType(_)) => ("type", None),
+                        Some(Pat(ptr)) => ("variable", Some(hash(ptr.syntax_node_ptr().range()))),
+                        Some(SelfParam(_)) => ("type", None),
+                        Some(GenericParam(_)) => ("type", None),
+                        None => ("text", None),
                     }
                 } else {
-                    "text"
+                    ("text", None)
                 }
             }
-            NAME => "function",
-            TYPE_ALIAS_DEF | TYPE_ARG | TYPE_PARAM => "type",
-            INT_NUMBER | FLOAT_NUMBER | CHAR | BYTE => "literal",
-            LIFETIME => "parameter",
-            T![unsafe] => "keyword.unsafe",
-            k if is_control_keyword(k) => "keyword.control",
-            k if k.is_keyword() => "keyword",
+            NAME => {
+                if let Some(name) = node.as_ast_node::<ast::Name>() {
```

Comment:
I think we should check if the parent of this name is `PatBinding`, an only then use `variable`, falling back to `function` overwise